src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1504 @@
+        ChipLogProgress(Zcl, "Location Location: Read SupportedLocations");
+        ReturnErrorOnFailure(ReadSupportedLocations(aEncoder));
+        break;
+    }
+
+    case ServiceArea::Attributes::SupportedMaps::Id: {
+        ChipLogProgress(Zcl, "Location Location: Read SupportedMaps");
+        ReturnErrorOnFailure(ReadSupportedMaps(aEncoder));
+        break;
+    } 
+
+    case ServiceArea::Attributes::SelectedLocations::Id: {
+        ChipLogProgress(Zcl, "Location Location: Read SelectedLocations");
+        ReturnErrorOnFailure(ReadSelectedLocations(aEncoder));
+        break;
+    } 
+
+    case ServiceArea::Attributes::CurrentLocation::Id: {
+        ChipLogProgress(Zcl, "Location Location: Read CurrentLocation");
+        ReturnErrorOnFailure(aEncoder.Encode(GetCurrentLocation()));
+        break;     
+    }
+
+    case ServiceArea::Attributes::EstimatedEndTime::Id: {
+        ChipLogProgress(Zcl, "Location Location: Read EstimatedEndTime");
+        ReturnErrorOnFailure(aEncoder.Encode(GetEstimatedEndTime()));
+        break;
+    }
+
+    case ServiceArea::Attributes::Progress::Id: {
+        ChipLogProgress(Zcl, "Location Location: Read Progress");
+        ReturnErrorOnFailure(ReadProgress(aEncoder));
+        break;
+    }
+
+    case ServiceArea::Attributes::FeatureMap::Id: {
+        ReturnErrorOnFailure(aEncoder.Encode(mFeature.Raw()));
+        break;
+    }
+  
+    default:
+        ChipLogProgress(Zcl, "Location Location: Read unsupported attribute %u", aPath.mAttributeId);
+        break;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    ChipLogProgress(Zcl, "%s", __func__); 
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case Commands::SelectLocations::Id:
+
+        CommandHandlerInterface::HandleCommand<Commands::SelectLocations::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSelectLocationsCmd(ctx, req); });
+        break;
+
+    case Commands::SkipCurrent::Id:
+
+        CommandHandlerInterface::HandleCommand<Commands::SkipCurrent::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSkipCurrentCmd(ctx); });
+        break;
+
+    default:
+        break;
+    }
+
+    return;
+}
+
+
+//*************************************************************************
+// attribute readers
+
+CHIP_ERROR Instance::ReadSupportedLocations(chip::app::AttributeValueEncoder & aEncoder)
+{
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        for (auto & entry : mSupportedLocations)
+        {
+            ReturnErrorOnFailure(encoder.Encode(entry.second));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR Instance::ReadSupportedMaps(chip::app::AttributeValueEncoder & aEncoder) 
+{
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
```

Comment:
Does the Matter client get a null value in case mSupportedMaps is empty?

Reply:
Copied this as the standard way to read a list. Tested just now - if supportedMaps is empty, it returns an empty list.

I found a number of nullable array attributes by grepping the cluster xmls, we need to add something like the following to return null when lists are empty. (as called for by the spec!)

        if (mSupportedMaps.size() == 0)
        {
            return aEncoder.EncodeNull();
        }
        else
        { 
           return aEncoder.EncodeList( ...